### ISSUE TITLE
Clustered VRP

### ISSUE THREAD

**Issue by N-Wouda** [PyVRP/PyVRP#470]
#### Is your feature request related to a problem? Please describe

This issue tracks clustered VRP and generalised VRP. The two are somewhat related, since they both specify clusters (groups) of clients. The difference is in what's required for these clusters:
- [ ] Clustered VRP requires all clients in the same group to be visited in sequence, one after another, before the route may continue with clients outside the cluster. We can already do this by modifying the distance matrix, but that does not work well with e.g. a max distance constraint.
- [x] Generalised VRP requires only a single client in the group is visited, rather than all of them. We already support optional clients, so we do not have to worry about not all locations being visited. But we do still need to build in some sort of support for this.

We do not have any facility in place to deal with such higher-level constraints currently.

#### Describe the solution you'd like

Support for clustered and generalised VRP.

#### Additional context

Came up in #441.

**Comment by N-Wouda** [comment]
Let's:
- [ ] Do a mini literature review on applications (relevance) and algorithmics (solution approaches);
- [ ] Think about an interface for this;
- [ ] Think about how this may be supported algorithmically.

**Comment by N-Wouda** [comment]
Interface could be something like:
```python
from pyvrp import Model

m = Model()

cluster = m.add_cluster(only_one=True/False, name="group")

client1 = m.add_client(..., cluster=cluster)  # now part of the cluster
client2 = m.add_client(..., cluster=None)  # default, not part of any cluster
```
One cannot add depots to clusters. Additionally, what should the default be for `only_one`? Should we provide a default? And is `only_one` a good name?

**Comment by N-Wouda** [comment]
OK, so some more thinking about the interface:
- I like `client_group` more than `cluster`. This also immediately makes clear that depots should not be passed - those are not clients!
- `only_one` could be renamed to `mutually_exclusive`, somewhat similar to how [argparse](https://docs.python.org/3/library/argparse.html#mutual-exclusion) handles mutual exclusion. I am not sure about a default value.

**Comment by N-Wouda** [comment]
The above definition also covers VRP with multiple time windows as a special case: 
- For each of a client's time windows, duplicate the client and assign it one of the time windows. 
- Mark the duplicated client as optional.
- Add all the duplicated clients to a mutually exclusive client group.
- Remove the original client with multiple time windows from the problem.
- ???
- Profit

**Comment by N-Wouda** [comment]
Depending on the speed at which we implement multi-trip VRP (and the speed at which I can finish my last paper 😆), this may or may not make the 0.8.0 release. Tentatively 0.9.0, but you never know!

**Comment by N-Wouda** [comment]
This description of the VRP with lunch breaks given in [Vidal et al. (2014)](https://doi.org/10.1016/j.ejor.2013.09.045) is simply a mutually exclusive client group that needs to be visited only when a route is non-empty:
<img width="497" alt="image" src="https://github.com/PyVRP/PyVRP/assets/16272507/aabcfbbf-e986-446b-8254-101530d2f134">
This will be really easy to add in if we support client groups.

**Comment by N-Wouda** [comment]
Returning to this, I have realised a few things:
- Clustered VRP should be possible already: simply make all edges into, and out of, the cluster sufficiently expensive. Then the solver should realise that entering and leaving the cluster only once (i.e., visit every client in the cluster consecutively) is optimal.
- Clustered VRP with weak cluster links - all clustered clients must be on the same route - is equivalent to a precedence constraint that requires all clustered clients to be visited before some dummy node is visited. So this is essentially a pickup and delivery problem, where all clustered clients want to route a zero load to the dummy node. When we tackle pickup and delivery in #331, we get this variant essentially for free. That seems to be a recurring theme.
- Generalised VRP is the only variant here that is non-trivial. So we should probably tailor this issue to focus only on that problem.

**Comment by N-Wouda** [comment]
> Returning to this, I have realised a few things:
> 
> * Clustered VRP should be possible already: simply make all edges into, and out of, the cluster sufficiently expensive. Then the solver should realise that entering and leaving the cluster only once (i.e., visit every client in the cluster consecutively) is optimal.
> * Clustered VRP with weak cluster links - all clustered clients must be on the same route - is equivalent to a precedence constraint that requires all clustered clients to be visited before some dummy node is visited. So this is essentially a pickup and delivery problem, where all clustered clients want to route a zero load to the dummy node. When we tackle pickup and delivery in [Pickup and delivery #331](https://github.com/PyVRP/PyVRP/issues/331), we get this variant essentially for free. That seems to be a recurring theme.
> * Generalised VRP is the only variant here that is non-trivial to express in the form of some other VRP. So we should probably tailor this issue to focus only on that problem.

Let's then add an `argparse` like interface to `pyvrp.Model`, since `argparse` solved this problem so well:
```python
m = pyvrp.Model()

group = m.add_mutually_exclusive_group()
client1 = group.add_client(...)
client2 = group.add_client(...)
```

**Comment by N-Wouda** [comment]
Moving this up to v0.8.0.

**Comment by N-Wouda** [comment]
We should:
- [ ] Find some benchmark instances;
- [ ] Think about the granular nieghbourhoods for mutually exclusive groups. They probably shouldn't be in each other's neighbourhood because only one of them can ever be in the solution at the same time.
- [ ] More?

**Comment by N-Wouda** [comment]
I'm waiting for a reply to a request for benchmark instances I emailed about yesterday. Once I have them I'll run a benchmark on literature instances.

**Comment by N-Wouda** [comment]
I have some benchmark instances for the GVRP now. With #514 nearly finished we can thus start benchmarking to compare.

**Comment by N-Wouda** [comment]
GVRP is now implemented. Let's get the benchmark instances underway and then we should be good to go for that variant. I'll keep the issue open for CluVRP.

**Comment by N-Wouda** [comment]
I'm running a benchmark of current main (c50ed1d335d3dba89d856a9095468a4467a6207e; job ID 7685011; 1h GVRP) on the benchmark instances of https://github.com/PyVRP/Instances/pull/10.

**Comment by N-Wouda** [comment]
GVRP is done; renaming this to CluVRP.

**Comment by N-Wouda** [comment]
The relevant part of this issue for 0.8.0 was GVRP, which is now done. CluVRP will probably happen at some point as well, but that's not currently planned. I suspect either in 0.9.0, or 0.10.0, but it's a bit of a stretch goal.

**Comment by N-Wouda** [comment]
Could be fun to pick this up for 0.10.0, to flesh out the client groups a bit more.

**Comment by N-Wouda** [comment]
Maybe combine with group-based service level (i.e., `min_num_services` for a group) based on what I gather from #441.

**Comment by N-Wouda** [comment]
There seems to be some interest in clustered VRP over at other solvers - https://github.com/reinterpretcat/vrp/issues/169.

**Comment by leonlan** [comment]
Leaving this here for reference, because I always get confused about the distinction between CluVRP/SoftCluVRP:

![Image](https://github.com/user-attachments/assets/2ff4de3f-8b63-4f54-ab81-4cdae778c7a1)

https://www.sciencedirect.com/science/article/pii/S0305054820302495

**Comment by N-Wouda** [comment]
Might possibly revisit this for 1.0.0; not sure. Depends a bit on which other features take priority in H2 2025.

**Comment by N-Wouda** [comment]
Discussed hard clusters with fixed sequence today. A special case of those are fixed prefix sequences like recently came up in #1011. Those are real easy because you can aggregate them prior to solving. We could support this pretty easily, I think.

**Comment by N-Wouda** [comment]
Will be part of the constraint system in PyVRP Enterprise.
